torch/nn/functional.py

```diff
@@ -4124,14 +4124,26 @@ def multi_head_attention_forward(query,                           # type: Tensor
 
     if key_padding_mask is not None:
         attn_output_weights = attn_output_weights.view(bsz, num_heads, tgt_len, src_len)
+        key_padding_mask = key_padding_mask.view(bsz, 1, 1, src_len)
         attn_output_weights = attn_output_weights.masked_fill(
-            key_padding_mask.unsqueeze(1).unsqueeze(2),
+            key_padding_mask,
             float('-inf'),
```

Comment:
I guess you can easily fix the problem if you set `float('-inf')` to `-1e-30`, right?

Reply:
This would probably under flow for half type

Reply:
Besides the potential issues with precision, wouldn't that also produce the uniform distribution under some use cases? These would probably be user errors, but it would be safer to mask correctly (or for the code to fail) than to silently produce the uniform distribution.

Reply:
Hi everyone,
- Recently, I've tried to fix & summarise this whole issue elaborately at https://github.com/pytorch/pytorch/issues/41508#issuecomment-1723119580.
- I did upload the **modified** basic scripts, which one can use readily for their experiments.
- And here I completely agree with @wgale and my comments, as well as changes completely inspired by his comments at #41508.

Thank You.